Samples/BackgroundBlur/BackgroundBlur/BackgroundBlur.cpp

```diff
@@ -0,0 +1,511 @@
+
+    MSG msg;
+
+    ZeroMemory(&msg, sizeof(msg));
+
+    // Perform application initialization.
+    if (!InitInstance(hInstance, nCmdShow))
+    {
+        NotifyError(NULL, L"Could not initialize the application.",
+            HRESULT_FROM_WIN32(GetLastError()));
+        return FALSE;
+    }
+
+    // Main message loop.
+    while (GetMessage(&msg, NULL, 0, 0))
+    {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    }
+
+    // Clean up.
+    if (g_pPlayer)
+    {
+        g_pPlayer->Shutdown();
+        SafeRelease(&g_pPlayer);
+    }
+    return 0;
+}
+
+//  Create the application window.
+BOOL InitInstance(HINSTANCE hInst, int nCmdShow)
+{
+    HWND hwnd;
+    WNDCLASSEX wcex;
+
+    g_hInstance = hInst; // Store the instance handle.
+
+    // Register the window class.
+    ZeroMemory(&wcex, sizeof(WNDCLASSEX));
+    wcex.cbSize = sizeof(WNDCLASSEX);
+    wcex.style = CS_HREDRAW | CS_VREDRAW;
+    wcex.lpfnWndProc = WndProc;
+    wcex.hInstance = hInst;
+    wcex.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
+    wcex.lpszMenuName = MAKEINTRESOURCE(IDC_MFPLAYBACK);
+    wcex.lpszClassName = szWindowClass;
+
+    if (RegisterClassEx(&wcex) == 0)
+    {
+        return FALSE;
+    }
+
+    // Create the application window.
+    hwnd = CreateWindow(szWindowClass, szTitle, WS_OVERLAPPEDWINDOW,
+        CW_USEDEFAULT, 0, CW_USEDEFAULT, 0, NULL, NULL, hInst, NULL);
+
+    if (hwnd == 0)
+    {
+        return FALSE;
+    }
+
+    ShowWindow(hwnd, nCmdShow);
+    UpdateWindow(hwnd);
+
+    return TRUE;
+}
+
+
+//  Message handler for the main window.
+
+LRESULT CALLBACK WndProc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    switch (message)
+    {
+    case WM_CREATE:
+        return OnCreateWindow(hwnd);
+
+    case WM_COMMAND:
+        switch (LOWORD(wParam))
+        {
+        case IDM_EXIT:
+            DestroyWindow(hwnd);
+            break;
+        case ID_FILE_OPENFILE:
+            OnFileOpen(hwnd);
+            break;
+        case ID_FILE_OPENURL:
+            OnOpenURL(hwnd);
+            break;
+        case ID_STREAM_START:
+            OnStartStream(hwnd);
+            break;
+        default:
+            return DefWindowProc(hwnd, message, wParam, lParam);
+        }
+        break;
+
+    case WM_PAINT:
+        OnPaint(hwnd);
+        break;
+
+    case WM_SIZE:
+        OnResize(LOWORD(lParam), HIWORD(lParam));
+        break;
+
+    case WM_ERASEBKGND:
+        // Suppress window erasing, to reduce flickering while the video is playing.
+        return 1;
+
+    case WM_DESTROY:
+        PostQuitMessage(0);
+        break;
+
+    case WM_CHAR:
+        OnKeyPress(wParam);
+        break;
+
+    case WM_APP_PLAYER_EVENT:
+        OnPlayerEvent(hwnd, wParam);
+        break;
+
+    default:
+        return DefWindowProc(hwnd, message, wParam, lParam);
+    }
+    return 0;
+}
+
+//  Open an audio/video file.
+void OnFileOpen(HWND hwnd)
+{
+    IFileOpenDialog* pFileOpen = NULL;
```

Comment:
should these be wrapped in smart pointers so you don't have to manually release the memory?

Reply:
This is all from demo code, but I can refactor to use smart pointers instead if it's the better coding practice to demonstrate.